File: netlogo-gui/src/main/shape/editor/DrawableList.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.shape.editor;

import org.nlogo.api.Shape;
import org.nlogo.api.ShapeList;
import org.nlogo.shape.ShapeChangeListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public strictfp class DrawableList
    extends javax.swing.JList<String>
    implements EditorDialog.VectorShapeContainer {
  javax.swing.DefaultListModel<String> listModel;
  final ShapeList shapeList;
  List<Shape> shapes;
  private final ShapeChangeListener shapeChangeListener;
  private java.awt.Component parent;

  public DrawableList(ShapeList shapeList, ShapeChangeListener listener,
               int rows, int height) {
    this.shapeChangeListener = listener;
    this.shapeList = shapeList;
    putClientProperty("Quaqua.List.style", "striped");
    setVisibleRowCount(rows);
    listModel = new javax.swing.DefaultListModel<String>();
    setModel(listModel);
    setFixedCellHeight(height);
  }

  public void setParent(java.awt.Component parent) {
    this.parent = parent;
  }

  //  Make sure the list of available shapes is up to date
  public void update() {
    listModel.clear();
    shapes = shapeList.getShapes();
    for (Shape shape : shapes) {
      listModel.addElement(shape.getName());
    }
  }

  public void update(Shape originalShape, Shape newShape) {
    // If you changed the name of the shape, get rid of the shape
    // with the old name
    if ((!originalShape.getName().equals(newShape.getName())) &&
        (!ShapeList.isDefaultShapeName(originalShape.getName()))) {
      removeShape(originalShape);
    }

    addShape(newShape);

    update(); // Update the shapes manager's list
    selectShapeName(newShape.getName());
  }

  // Select a shape in the list
  public void selectShapeName(String name) {
    int index = -1;
    while (++index < listModel.size())    // Iterate through all the items of the list until you have
    {                    //  the index of <name>
      if (listModel.elementAt(index).equals(name)) {
        break;
      }
    }
    addSelectionInterval(index, index);    // Select that index
    ensureIndexIsVisible(index);
  }

  public Object elementAt(int index) {
    return listModel.elementAt(index);
  }

  public Shape getOneSelected() {
    int[] selected = getSelectedIndices();
    if (selected.length == 1)   // You can only edit one shape at a time
    {
      return shapes.get(selected[0]);
    }

    return null;
  }

  public Shape getShape(int index) {
    return shapes.get(index);
  }

  public Set<String> getShapeNames() {
    return shapeList.getNames();
  }

  public boolean exists(String name) {
    return shapeList.exists(name);
  }

  // Select the shape with a given index
  void selectShapeIndex(int index) {
    addSelectionInterval(index, index);
    ensureIndexIsVisible(index);
  }

  // Delete a shape of the current model
  List<Shape> deleteShapes() {
    int[] selected = getSelectedIndices();
    int delete;
    List<Shape> deletedShapes = new ArrayList<Shape>();

    // Confirm that the user wants to delete
    if (selected.length > 1) {
      delete = javax.swing.JOptionPane.showConfirmDialog
          (parent, "Are you sure you want to delete these "
              + selected.length + " shapes?", "Delete", javax.swing.JOptionPane.YES_NO_OPTION);
    } else if (selected.length == 1) {
      delete = javax.swing.JOptionPane.showConfirmDialog(parent, "Are you sure you want to delete this shape?",
          "Delete", javax.swing.JOptionPane.YES_NO_OPTION);
    } else {
      return deletedShapes;
    }

    if (delete != javax.swing.JOptionPane.YES_OPTION) {
      return deletedShapes;
    }

    for (int i = 0; i < selected.length; ++i)      // Remove the selected shapes from the model
    {
      // Don't delete the default turtle
      if (!ShapeList.isDefaultShapeName(shapes.get(selected[i]).getName())) {
        Shape shape = shapes.get(selected[i]);
        deletedShapes.add(shape);
        removeShape(shape);
      }
    }

    for (int i = selected.length - 1; i >= 0; --i)    // Update the shapes manager to reflect those deletions
    {
      if (!(ShapeList.isDefaultShapeName(shapes.get(selected[i]).getName()))) {
        shapes.remove(selected[i]);
        listModel.remove(selected[i]);
      }
    }

    selectShapeIndex(((selected[0] >= shapes.size()) ? selected[0] - 1 : selected[0]));
    return deletedShapes;
  }

  void addShape(Shape shape) {
    if (shape != null) {
      Shape replacedShape = shapeList.add(shape);
      if (shapeChangeListener != null) {
        shapeChangeListener.shapeChanged
            (replacedShape == null ? shape : replacedShape);
      }
    }
  }

  void removeShape(Shape shape) {
    Shape removedShape = shapeList.removeShape(shape);
    if (removedShape != null && shapeChangeListener != null) {
      shapeChangeListener.shapeRemoved(shape);
      shapeChangeListener.shapeChanged(removedShape);
    }
  }

}

File: netlogo-gui/src/main/shape/editor/DrawableList.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.shape.editor

import org.nlogo.core.{ I18N, Shape, ShapeList, ShapeListTracker }
import org.nlogo.swing.OptionPane

import scala.reflect.ClassTag

class DrawableList[A <: Shape](shapeTracker: ShapeListTracker, rows: Int, height: Int, parent: java.awt.Component)(implicit ct: ClassTag[A])
  extends javax.swing.JList[A]
  with EditorDialog.VectorShapeContainer {

  def shapeList = shapeTracker.shapeList

  val listModel = new javax.swing.DefaultListModel[A]()

  setVisibleRowCount(rows)
  setModel(listModel)
  setFixedCellHeight(height)
  setCellRenderer(new ShapeCellRenderer())

  //  Make sure the list of available shapes is up to date, filtering by name if provided
  def update(name: Option[String] = None): Unit = {
    listModel.clear()

    name match {
      case Some(str) =>
        shapeList.shapes.foreach {
          case s: A if s.name.toLowerCase.contains(str.toLowerCase) => listModel.addElement(s)
          case _ =>
        }

      case _ =>
        shapeList.shapes.foreach {
          case s: A => listModel.addElement(s)
          case _ =>
        }
    }
  }

  override def update(originalShape: Shape, newShape: Shape): Unit = {
    // If you changed the name of the shape, get rid of the shape
    // with the old name
    if (originalShape.name != newShape.name && ShapeList.DefaultShapeName != originalShape.name)
      removeShape(originalShape)

    addShape(newShape)

    update() // Update the shapes manager's list
    selectShapeName(newShape.name)
  }

  // Select a shape in the list
  def selectShapeName(name: String): Unit = {
    var index = 0
    var namedIndex = -1
    // Iterate through items to find index of <name>
    while (index < listModel.size && namedIndex == -1) {
      if (listModel.getElementAt(index).name == name)
        namedIndex = index
      index += 1
    }
    addSelectionInterval(namedIndex, namedIndex) // Select that index
    ensureIndexIsVisible(namedIndex)
  }

  def elementAt(index: Int): A =
    listModel.getElementAt(index)

  def getOneSelected: Option[A] = {
    val selected: Array[Int] = getSelectedIndices
    if (selected.length == 1)
      Option(listModel.get(selected(0)))
    else
      None
  }

  def getShape(index: Int): Option[A] =
    Option(listModel.get(index))

  def getShapeNames: Set[String] =
    shapeList.names

  def exists(name: String): Boolean =
    shapeList.exists(name)

  // Select the shape with a given index
  def selectShapeIndex(index: Int): Unit = {
    if (index < listModel.size) {
      addSelectionInterval(index, index)
      ensureIndexIsVisible(index)
    }
  }

  private def confirmDeletion(deleteCount: Int): Boolean = {
    val message =
      if (deleteCount > 1)
        Some(I18N.gui.getN("tools.shapesEditor.delete.many.confirm", Int.box(deleteCount)))
      else if (deleteCount == 1)
        Some(I18N.gui.get("tools.shapesEditor.delete.one.confirm"))
      else
        None

    message.map { msg =>
      new OptionPane(parent, I18N.gui.get("tools.shapesEditor.delete"), msg, OptionPane.Options.YesNo,
                     OptionPane.Icons.Question).getSelectedIndex == 0
    }.getOrElse(false)
  }

  // Delete a shape of the current model
  def deleteShapes(): Seq[Shape] = {
    val selected: Array[Int] = getSelectedIndices

    if (confirmDeletion(selected.size)) {
      // Remove the selected shapes from the model
      val shapesToDelete = selected
        .map(i => (i, listModel.get(i)))
        .filter(_._2.name != ShapeList.DefaultShapeName)

      // we have to reverse here to delete indices in
      // decreasing order
      shapesToDelete.reverse.foreach {
        case (i, shape) =>
          removeShape(shape)
          listModel.remove(i)
      }

      val newSelectedShapeIndex =
        if (selected(0) >= listModel.size)
          selected(0) - 1
        else
          selected(0)

      selectShapeIndex(newSelectedShapeIndex)
      shapesToDelete.collect {
        case (_, s: A) => s
      }.toSeq
    } else {
      Seq()
    }
  }

  def addShape(shape: Shape): Unit = {
    Option(shape).foreach(shapeTracker.add)
  }

  def removeShape(shape: Shape): Unit = {
    Option(shape).foreach(shapeTracker.removeShape)
  }
}
